src/ring_theory/algebra.lean

```diff
@@ -576,10 +576,38 @@ def module.restrict_scalars : module R E :=
   add_smul  := by simp [add_smul],
   zero_smul := by simp [zero_smul] }
 
+/--
+The identity function, considered as an `R`-linear map from an `R`-algebra `S` to itself,
+with `module.restrict_scalars R S S` as the module structure in the source,
+and `algebra.to_module` as the module structure in the target.
+
+Unfortunately these structures are not generally definitionally equal,
+so we sometimes need to insert this map in order to typecheck.
+-/
+def algebra.restrict_scalars_iso :
```

Comment:
I imagine you could go back and forth between the two structures, so you should rather construct the linear equiv between the two structures.

Reply:
We'll have to merge #2444 first, otherwise you can't even state 
```lean
@[simp]
lemma algebra.restrict_scalars_equiv_symm_apply (s : S) :
  (algebra.restrict_scalars_equiv R S).symm s = s := rfl
```
without the dreaded:
```
synthesized type class instance is not definitionally equal to expression inferred by typing rules, synthesized
  algebra.to_module
inferred
  module.restrict_scalars R S S
```

Reply:
I've merge #2444 into this, so this is now blocked on that, but otherwise I've made the changes you suggested.